sisl/geom/nanoribbon.py

```diff
@@ -154,3 +159,461 @@ def zgnr(width, bond=1.42, atoms=None):
+                situation, an error will be raised.
+
+                If not provided, it defaults to 1.
+            shift: int, optional
+                The shift of this section with respect to the previous one.
+
+                The exact meaning of the value depends on the sections that are
+                being joined so that only valid ribbons are produced:
+                    - If both ribbons are odd, ribbons are aligned on the center
+                    and then shifted:
+                        ^ (shift) zig zag atoms if the incoming section is thinner than the
+                            previous one.
+                        ^ (2 * shift) zig zag atoms if they differ in a multiple of 4 atoms
+                            and both configurations are closed or if they differ in (2 + a
+                            multiple of 4 atoms) and configurations are different (one open
+                            and one closed.)
+                        ^ (1 + 2 * shift) zig zag atoms otherwise.
+                    - If both ribbons are even, they are aligned by the open end of the last
+                    section, not the incoming one, and then they are shifted:
+                        ^ (0) zig zag atoms if the shift is 0. In this case both sections are 
+                        forced to have the same edge open.
+                        ^ (-1 + 2 * abs(shift)) zig zag atoms in the appropiate direction if abs(shift) is
+                        bigger than 1. Notice that you can only shift the open edges toward the center.
+                    - If the ribbons have different parities, they are aligned on the open edge
+                    of the even ribbon and then shifted:
+                        ^ (0) zig zag atoms if shift is 0 in the particular cases that (i) the incoming
+                        odd section is thinner than the last even section or (ii) the last section, 
+                        being odd, has a border in the open configuration. Only in those cases does it make
+                        sense to exactly match the edges.
+                        ^ (bias + 2 * abs(shift)) zig zag atoms in the appropiate direction, where `bias` is
+                        either -1 or 1 depending on the exact situation and always ensures that the minimum
+                        shift is always `abs(shift) == 1`.
+
+                If not provided, it defaults to 0.
+            atoms: Atom, optional
+                Value to pass to the `atoms` argument of `nanoribbon`. If not provided,
+                it defaults to the `atoms` argument passed to this function.
+    bond: float
+        The bond length of the ribbon.
+    atoms: Atom, optional
+        The default atom to use for the sections. For each section, if it contains an `atoms`
+        value, it will be used instead of this one.
+    kind: {'armchair', 'zigzag'}, optional
+        The kind of ribbon to be produced. Defaults to 'armchair'.
+
```

Comment:
I would like to see an `Examples` section that shows the kind of structures with the shifts, that would be cool. Perhaps very narrow ones that can be shown easily in an ASCII drawing?

Reply:
I will add examples.

Do you know how to draw a ribbon in ASCII without using all your lifetime? I would rather not do it, but if you want to do it feel free :laughing:.

Idea for the future: ASCII backend for `GeometryPlot`.

Reply:
Hehe, yeah, it was just an idea. :)

Reply:
I don't know if there exists an easy ascii molecular drawer

Reply:
`jp2a` is a quite nice command line tool to convert from jpg to ascii. Here is an example:
```python
composite_agnr([5, 6, 7, 7, 6, 5]).plot(axes="xy", show_cell=False, show_atoms=True, atoms_scale=0.1)
```
![newplot-1](https://user-images.githubusercontent.com/11407069/149306388-85121523-e473-4d3b-ad48-cd054a17650a.png)
Then I manipulate with `gimp` to remove the frame and to get thick black bonds, and to export in `jpg` format:
![structure_fat](https://user-images.githubusercontent.com/11407069/149305839-bfa28b90-e594-4ab2-a8c9-8792caf3b99b.jpg)
Then it can be converted to ascii:
```
$ jp2a --chars=****"  " --width=80 -b structure.jpg
+--------------------------------------------------------------------------------+
|                                   **********                                   |
|                                  **        **                                  |
|                                ***          ***                                |
|                        **********            **********                        |
|                      ***      ***            ***      ***                      |
|                     **          **          **          **                     |
|                   ***            ***      ***            ***                   |
|          ************            ************            ************          |
|         ***        ***          ***        ***          ***        ***         |
|        **            **        ***           **        **            **        |
|**********            ************            ************            **********|
|        ***          ***        ***          ***        ***          ***        |
|         ***        ***          ***        ***          ***        ***         |
|          ************            ************            ************          |
|          ***      ***            ***       **            **        **          |
|         **          **          **          **          **          **         |
|**********            ************            ************            **********|
|**********            ************            ************            **********|
|        ***          ***        ***          ***        ***          ***        |
|          **        **            **        **            **        **          |
|           **********              **********              **********           |
+--------------------------------------------------------------------------------+
```
It requires a bit of playing around to get a reasonably looking result. But maybe this could be automatized?

Reply:
The question is, would someone use it? Hahah

I've seen that `gnuplot` can draw in ascii and there is an extension for `matplotlib` (https://github.com/gooofy/drawilleplot) that can do it too.

Reply:
Yeah, this is overkill here. :)